### PR TITLE
Add drop rows with negative or zero timepoints

### PR DESCRIPTION
**Pull request by zsewa**
This PR adds checks for negative and zero values for the timestamps in the preprocessing of the data.
Positive values are needed for the log-transformation of the survival times.

Bad values are handled by dropping the offending rows.
The number of dropped rows are logged and published to the controller by setting an error state including the split name.

### CODE REVIEW

File: engine/survival_svm.py [hunk 1]
```diff
@@ -247,6 +247,19 @@ def run(self):
...
+        split_manager = self.app.internal.get('split_manager')
+        for split in split_manager:
+            if not split.data.get('opt_out'):
+                model: LocalTraining = split.data.get('model')
+
+                dropped_rows = model.data.drop_negative_and_zero_timepoints()
+                if dropped_rows > 0:
+                    self.update(
+                        message=f'Dropped {dropped_rows} samples with zero or negative timepoints in {split.name}',
+                        state=STATE_ERROR
```
Comment (julianspaeth) [inline engine/survival_svm.py:260]: I think we should use the "warning" state here. The error indicates that the app can not successfully finish. But here, we take care of the problems, which would make a warning more appropriate.

File: logic/data.py [hunk 2]
```diff
@@ -53,6 +53,23 @@ def __init__(self, X: pd.DataFrame, y: NDArray):
             time_to_event=time_to_event,
         )
 
+    def drop_negative_and_zero_timepoints(self) -> int:
```
Comment (julianspaeth) [inline logic/data.py:56]: Besides dropping the negative and zero time points, we also wanted to add functionality to the config file to decide the event value and censored value. Could you add this check @zsewa?

File: logic/data.py [hunk 3]
```diff
@@ -53,6 +53,23 @@ def __init__(self, X: pd.DataFrame, y: NDArray):
...
+        """
+        Drop rows with a negative or zero time to prepare data for log-transformation.
+
+        :return: Number of dropped rows
+        """
+        logging.info("Check data for negative and zero timepoints")
+        bad_rows = self.survival.time_to_event <= 0
```
Comment (julianspaeth) [inline logic/data.py:63]: I think this would keep the rows where the survival time is e.g. a string.

**Review by julianspaeth — CHANGES_REQUESTED** [review]

File: engine/survival_svm.py (hunk 1 above)
Comment (zsewa) [reply engine/survival_svm.py:260]: Not sure how to set the warning state, as https://featurecloud.ai/assets/api/redoc-static.html#operation/status lists the status as an enum of the states `running`, `error`, `action_required`.
Is there an extra `warning` state or should I keep it in the `running` state than?

File: logic/data.py (hunk 3 above)
Comment (zsewa) [reply logic/data.py:63]: Non-numeric values should not be possible at this point. I suspect that text values that can not be interpreted as a numeric value will raise a `ValueError` in `logic.data.read_survival_data_np`. However, I don't have the test setup right now to verify this. Could you start a run with non-numeric values for the time, please?
Anyway, the preprocessing steps are a bit all over the place. I think I need to revise the data preprocessing part a bit to allow for good warning messages here.

File: engine/survival_svm.py (hunk 1 above)
Comment (julianspaeth) [reply engine/survival_svm.py:260]: I see. I suggest to keep it in running then.

File: logic/data.py (hunk 3 above)
Comment (julianspaeth) [reply logic/data.py:68]: I will check it later, I think for now this is fine

File: logic/data.py (hunk 2 above)
Comment (zsewa) [reply logic/data.py:63]: I will open another PR for this momentarily.

**Pull request merged**